fix(route53-targets): A/AAAA Alias Record to ELB cannot resolve IPv6 addresses

BryanPan342 opened the pull request:
**[ISSUE]**
Route53 A/AAAA Alias Record created by CDK didn't add prefix 'dualstack' to the dnsName of ElasticLoadBalancer as suggested in [documentation](https://aws.amazon.com/de/premiumsupport/knowledge-center/alias-resource-record-set-route53-cli/), resulting in inability to resolve IPv6 records (outcome consistent with these [docs](https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/elb-internet-facing-load-balancers.html)).

**[APPROACH]**
Joined prefix: _dualstack_ to DNS of ELB targets to resolve IPv6 records. 

Example:
> If ELB DNS name is `ALB-xxxxxxxx.us-west-2.elb.amazonaws.com`,
> Route53 A/AAAA Alias Record will create Alias Target `dualstack.ALB-xxxxxxxx.us-west-2.elb.amazonaws.com`

**[NOTE]** 
Ignoring the case of using `ApiGatewayDomain(domainName)` because the prefix of 'dualstack' should be added in the `domainName` parameter.

```
new route53.ARecord(this, 'AliasRecord', {
  zone,
  target: route53.RecordTarget.fromAlias(new alias.ApiGatewayDomain(domainName)),
});
```

If creating an ApiGatewayDomain for ELB, endpoint for ELB with prefix 'dualstack' should be declared there, not within route53 package.

Fixes #6271 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 5d38010e62a24a2e1ec8fd96170ca0ececebc5f6
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5ea8c22e-1c9d-4540-bc2a-742f84cdb435%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b415cd0187cbc33dd2b0639de3b530829285c65d
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e44c6efb-1426-4095-8e51-08da5de609e8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

shivlaks commented on packages/@aws-cdk/aws-route53-targets/lib/classic-load-balancer-target.ts:
```diff
@@ -9,9 +10,11 @@ export class ClassicLoadBalancerTarget implements route53.IAliasRecordTarget {
     return {
       hostedZoneId: this.loadBalancer.loadBalancerCanonicalHostedZoneNameId,
-      dnsName: this.loadBalancer.loadBalancerDnsName,
+      dnsName: cdk.Fn.join(delimiter, prefix.concat([this.loadBalancer.loadBalancerDnsName])),
```
```suggestion
      dnsName: cdk.Fn.join('.', ['dualstack', this.loadBalancer.loadBalancerDnsName]),
```

shivlaks commented on packages/@aws-cdk/aws-route53-targets/lib/load-balancer-target.ts:
```diff
@@ -9,9 +10,11 @@ export class LoadBalancerTarget implements route53.IAliasRecordTarget {
   }
 
   public bind(_record: route53.IRecordSet): route53.AliasRecordTargetConfig {
+    const delimiter: string = '.';
+    const prefix: string[] = ['dualstack'];
     return {
       hostedZoneId: this.loadBalancer.loadBalancerCanonicalHostedZoneId,
-      dnsName: this.loadBalancer.loadBalancerDnsName,
+      dnsName: cdk.Fn.join(delimiter, prefix.concat([this.loadBalancer.loadBalancerDnsName])),
```
same changes as suggested in classic

BryanPan342 commented on yarn.lock:
```diff
@@ -1887,6 +1887,11 @@ anymatch@^3.0.3:
     normalize-path "^3.0.0"
     picomatch "^2.0.4"
 
+app-root-path@^2.2.1:
```
oh my gosh, I knew it was a dependency bug but didn't think it was that easy 😭

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 19ce0d491db946b2f009f58a65570086c8617e17
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ee6a727c-d379-4c64-b613-1a7a99d9a799%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 8f720206d0319b0a1ace5d9ff9db6733d084ee58
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9b630d57-cec9-4bed-b140-c33c536835e7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b68f213217fde7fa4cf6037fdf812aa531c80451
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d7728df1-7e22-425d-93bb-f5cdebfae821%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 0ccfa93a30a159cf01f119640fad9d5c28437a69
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=51e154c6-b680-4b29-8dc6-e9a6b080b4e5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c2d023612c6d271c16f74a1c125d9a44fe95d5e2
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=dcf55f9c-f600-49a7-9568-1ea3331f2a9c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jogold commented on packages/@aws-cdk/aws-route53-targets/lib/load-balancer-target.ts:
```diff
@@ -11,7 +12,7 @@ export class LoadBalancerTarget implements route53.IAliasRecordTarget {
     return {
       hostedZoneId: this.loadBalancer.loadBalancerCanonicalHostedZoneId,
-      dnsName: this.loadBalancer.loadBalancerDnsName,
+      dnsName: cdk.Fn.join('.', ['dualstack', this.loadBalancer.loadBalancerDnsName]),
```
Why not ``dualstack.${this.loadBalancer.loadBalancerDnsName}`` and let `core` handle the case when a `Fn.join` is effectively needed.

shivlaks replied:
even better, let's go with that!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 4436fc57181957c70614cf5e8f51d11c0c3c304f
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2877e27b-2456-456b-83ad-74b41c6ce61c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: e0055432d29bde3c4126ff60aa4f89bff56fef79
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b2ba69eb-6de8-4bc1-87ab-01feda2e56cc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

BryanPan342 commented:
> nice job getting this one started!!!
> 
> I have a couple of suggestions/questions:
> 
> 1. Confirm this is not a breaking change
> 
> * We are updating an existing property here and we want to make sure we don't break customers as this is a stable module. From the [documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-route53-recordset.html#aws-properties-route53-recordset-properties), it looks like update to `AliasTarget` requires no interruption and so it should be safe to introduce. Worth verifying that the documentation is accurate.
> 
> 1. Commit Message & Body
> 
> * In your commit description, outline the root cause of the issue and describe the approach taken to fix, any alternatives, considerations, caveats, gotchas, etc. You don't need the code snippets in the commit body
> * change the commit message to describe the issue. The reason for this is because this will show up in the [release notes](https://github.com/aws/aws-cdk/releases/tag/v1.47.0) as bug fixes. It's easier for users to read it when the message describes the bug that was fixed. In this case it might be something like ` A/AAAA Alias Record to ELB cannot resolve IPv4 and IPv6 addresses`
> 
> 1. Update the README.
> 
> * It's a little plain right now, but let's start improving the README with changes that are in scope for this fix. That might simply be adding some documentation links and describing the behaviour since that's changing now.
> 
> I believe you also have some tests in other modules that use the ELB target in their tests need to be updated

no longer a breaking change ✅
updated commit message ✅
updated readme ✅
